### PR TITLE
Use encoder

### PR DESCRIPTION
**Pull request by GuillaumeDesforges**
This PR proposes a shift in the way we ingest and process DICOM data in Spark.

The first approach (using data source API) wanted to achieve a mapping between DICOM structure and Spark internal row structure, by turning a DICOM binary representation (from input file) into a Spark Row, associated with a schema.

My research showed me that this first approach has many drawbacks:
* parsing DICOM to a Spark structured schema in a generic way is far from trivial, as DICOM can have infinitely deep nested datasets, which would be messy to write to Spark Rows thus making it hard to manipulate for the user
* it would be hard to make changes after the first implementation as it would each time require a (costly) migration

On the other hand, DICOM has always been serialized to/deserialized from binary data, which Spark can easily store and pass around.
The new approach would be to store this DICOM data as binary in Spark internal row representation, but allow users to manipulate it with a proper API, while serialization should be "hidden" to the user.

One can indeed use `Dataset[U : Encoder]`, which makes you manipulate it with function like `Dataset[U : Encoder].map[V: Encoder](U => V)` (which of course specializes partially to `Dataset[U : Encoder].map(U => U)`

Most primitives or useful type already have encoders implemented in spark. One can also make encoders of product types using `org.apache.spark.sql.Encoders.product[P]`, or encoders for serializables using `org.apache.spark.sql.Encoders.kryo[S]`.

However, there is currently no evident user API to allow one to make their own encoders.
From exploring the Encoder API, we can understand that `Encoder` is a trait which is matched on by Spark in the evaluation of a task, with currently only one implementation supported: `ExpressionEncoder`, which is a Catalyst object to generate JVM code.

In conclusion, providing such an encoder as an implicit value in a context where spark is used should be sufficient to manipulating Dicom as an object, while its data rests in Spark in the form of one column binary data.

### CODE REVIEW

**Review by robopoc — COMMENTED** [review]
I think this is now close to just using RDDs. Let's discuss in a bit....

**Comment by GuillaumeDesforges** [comment]
Closed in favor of using the Data Source API
